tools/build_tar/buildtar.go

```diff
@@ -330,7 +329,105 @@ func (f *tarFile) addTar(toAdd string) error {
 }
 
 func (f *tarFile) addDeb(toAdd string) error {
-	return fmt.Errorf("addDeb unimplemented")
+	pkgDataFound, pkgMetadataFound := false, false
+	var r io.Reader
+	file, err := os.Open(toAdd)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	r = file
+	r = bufio.NewReader(r)
+	tr := ar.NewReader(r)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		parts := strings.SplitN(header.Name, ".", 2)
+		name := strings.TrimRight(header.Name, "/")
+		switch parts[0] {
+		case "data":
+			pkgDataFound = true
+			err = f.addTar(tr, name)
+			if err != nil {
+				return err
+			}
+		case "control":
+			pkgMetadataFound = true
+			f.addPkgMetadata(tr, name)
+		}
+	}
+	if !pkgDataFound {
+		return fmt.Errorf("%s does not contain a data file", toAdd)
+	}
+	if !pkgMetadataFound {
+		return fmt.Errorf("%s does not contain a control file", toAdd)
+	}
+	return nil
+}
+
+func (f *tarFile) addPkgMetadata(r io.Reader, toAdd string) error {
+	r, err := decompress(r, toAdd)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(r)
+	controlFileFound := false
+	var buf bytes.Buffer
+	var header *tar.Header
+	for {
+		header, err = tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if filepath.Base(header.Name) == pkgMetadataFile {
+			controlFileFound = true
+			io.Copy(&buf, r)
+			break
+		}
+
+	}
+	if !controlFileFound {
+		return fmt.Errorf("%s does not contain control file", toAdd)
+	}
+	metadata := buf.String()
+	destinationFile := filepath.Join(dpkgStatusDir, parsePkgName(metadata, toAdd))
+	header.Name = destinationFile
+	if err := f.tw.WriteHeader(header); err != nil {
+		return err
+	}
+	if _, err := f.tw.Write(buf.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func decompress(r io.Reader, fName string) (io.Reader, error) {
+	switch {
+	case strings.HasSuffix(fName, "gz"):
+		return gzip.NewReader(r)
+	case strings.HasSuffix(fName, "bz2"):
+		return bzip2.NewReader(r), nil
+	case strings.HasSuffix(fName, "xz"):
+		return xz.NewReader(r)
+	default:
+		return r, nil
+	}
+}
+
+func parsePkgName(metadata, filename string) string {
```

Comment:
a brief docstring here describing this would be good, too.

Reply:
fixed name to be more descriptive